apollo-api/src/commonMain/kotlin/com/apollographql/apollo3/api/Version2CustomTypeAdapterToAdapter.kt

```diff
@@ -0,0 +1,73 @@
+package com.apollographql.apollo3.api
+
+import com.apollographql.apollo3.api.json.JsonReader
+import com.apollographql.apollo3.api.json.JsonWriter
+import kotlin.jvm.JvmStatic
+
+/**
+ * An [Adapter] that wraps an Apollo Android v2 style [CustomTypeAdapter], to ease migration from v2 to v3.
+ */
+@Deprecated("Used for backward compatibility with 2.x, use Adapter instead")
+class Version2CustomTypeAdapterToAdapter<T>(
+    private val v2CustomTypeAdapter: CustomTypeAdapter<T>,
+) : Adapter<T> {
+  override fun fromJson(reader: JsonReader, customScalarAdapters: CustomScalarAdapters): T {
+    val value: Any? = if (reader.peek() == JsonReader.Token.NULL) null else AnyAdapter.fromJson(reader)
+    return v2CustomTypeAdapter.decode(CustomTypeValue(value))
+  }
+
+  override fun toJson(writer: JsonWriter, customScalarAdapters: CustomScalarAdapters, value: T) {
+    val encoded: Any? = v2CustomTypeAdapter.encode(value).value
+    if (encoded == null) {
+      writer.nullValue()
+    } else {
+      AnyAdapter.toJson(writer, encoded)
+    }
+  }
+}
+
+/**
+ * A replica of Apollo Android v2's CustomTypeAdapter, to ease migration from v2 to v3.
+ *
+ * Make your CustomTypeAdapters implement this interface by updating the imports
+ * from `com.apollographql.apollo.api` to `com.apollographql.apollo3.api`
+ * and wrap them in [Version2CustomTypeAdapterToAdapter]s, which you can add to
+ * `ApolloClient.Builder` or [CustomScalarAdapters].
+ */
+@Deprecated("Used for backward compatibility with 2.x, use Adapter instead")
+interface CustomTypeAdapter<T> {
+  fun decode(value: CustomTypeValue<*>): T
+  fun encode(value: T): CustomTypeValue<*>
+}
+
+/**
+ * A replica of Apollo Android v2's CustomTypeValue, to ease migration from v2 to v3.
+ *
+ * In your [CustomTypeAdapter], update the imports from `com.apollographql.apollo.api` to
+ * `com.apollographql.apollo3.api` to use this version.
+ */
+@Deprecated("Used for backward compatibility with 2.x, use Adapter instead")
+open class CustomTypeValue<T>(val value: T) {
+  object GraphQLNull : CustomTypeValue<Unit>(Unit)
+  class GraphQLString(value: String) : CustomTypeValue<String>(value)
```

Comment:
These used to implement `equals`/`hashcode`. I don't think it's required but asking just in case, is it important?

Reply:
Good question! I stripped them as a simplification but without too much thinking 😅. Looking at it more a bit more closely at how they are used, I don't think this would break anything.

Reply:
Sounds good 👍